addons/controllers/clusterbootstrap_controller.go

```diff
@@ -1361,3 +1388,107 @@ func (r *ClusterBootstrapReconciler) reconcileClusterProxyAndNetworkSettings(clu
 
 	return nil
 }
+
+func (r *ClusterBootstrapReconciler) reconcileDelete(cluster *clusterapiv1beta1.Cluster, log logr.Logger) (ctrl.Result, error) {
+	remoteClient, err := util.GetClusterClient(r.context, r.Client, r.Scheme, clusterapiutil.ObjectKey(cluster))
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	err = remoteClient.Get(r.context, client.ObjectKeyFromObject(cluster), clusterBootstrap)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if clusterBootstrap == nil {
+		return ctrl.Result{}, nil
+	}
+
+	err = r.removeAdditionalPackageInstalls(cluster, log)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	packageInstallsArePresent := hasPackageInstalls(r.context, remoteClient, cluster, r.Config.SystemNamespace, clusterBootstrap.Spec.AdditionalPackages, log)
+	timeOutReached := time.Now().After(cluster.GetDeletionTimestamp().Add(clusterDeleteTimeout))
+
+	if packageInstallsArePresent || timeOutReached {
+		return ctrl.Result{RequeueAfter: clusterDeleteRequeTime}, nil
+	}
+	controllerutil.RemoveFinalizer(clusterBootstrap, types.AddonFinalizer)
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		return r.Client.Update(r.context, clusterBootstrap)
+	})
+	if err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	controllerutil.RemoveFinalizer(cluster, types.AddonFinalizer)
+	err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		return r.Client.Update(r.context, cluster)
+	})
+	if err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func hasPackageInstalls(ctx context.Context, remoteClient client.Client,
+	cluster *clusterapiv1beta1.Cluster, namespace string, packages []*runtanzuv1alpha3.ClusterBootstrapPackage,
+	log logr.Logger) bool {
+
+	for _, pkg := range packages {
+		pkgInstallName := util.GeneratePackageInstallName(cluster.Name, pkg.RefName)
+		if packageInstallExists(ctx, pkgInstallName, namespace, remoteClient, log) {
+			return true
+		}
+	}
+	return false
+}
+
+func packageInstallExists(ctx context.Context, pkgInstallName, pkgInstallNamespace string,
+	remoteClient client.Client, log logr.Logger) bool {
+
+	pkgInstall := &kapppkgiv1alpha1.PackageInstall{}
+	if err := remoteClient.Get(ctx, client.ObjectKey{Name: pkgInstallName, Namespace: pkgInstallNamespace}, pkgInstall); err != nil {
+		if apierrors.IsNotFound(err) {
+			return false
+		}
+		log.Error(err, "Could not verify if package install is present")
+		return true
+	}
+	return true
+}
+
```

Comment:
perhaps there is something in https://github.com/vmware-tanzu/tanzu-framework/pull/1781 we can use instead of packagesInstallExists and hasPackageInstalls

Reply:
investigated and did not find anything we could use here